app/src/org/commcare/activities/connect/ConnectIDDatabaseHelper.java

```diff
@@ -0,0 +1,119 @@
+package org.commcare.activities.connect;
+
+import android.content.Context;
+import android.os.Build;
+
+import net.sqlcipher.database.SQLiteDatabase;
+
+import org.commcare.CommCareApplication;
+import org.commcare.android.database.connect.models.ConnectLinkedAppRecord;
+import org.commcare.android.database.connect.models.ConnectUserRecord;
+import org.commcare.android.database.global.models.ApplicationRecord;
+import org.commcare.android.database.global.models.ConnectKeyRecord;
+import org.commcare.models.database.AndroidDbHelper;
+import org.commcare.models.database.SqlStorage;
+import org.commcare.models.database.connect.DatabaseConnectOpenHelper;
+import org.commcare.modern.database.Table;
+import org.commcare.utils.EncryptionUtils;
+import org.javarosa.core.services.Logger;
+import org.javarosa.core.services.storage.Persistable;
+
+import java.util.Vector;
+
+public class ConnectIDDatabaseHelper {
+    private static final Object connectDbHandleLock = new Object();
+    private static SQLiteDatabase connectDatabase;
+
+    private static byte[] getConnectDBPassphrase(Context context) {
+        try {
+            for (ConnectKeyRecord r : CommCareApplication.instance().getGlobalStorage(ConnectKeyRecord.class)) {
+                return EncryptionUtils.decryptFromBase64String(context, r.getEncryptedPassphrase());
```

Comment:
This crashes for me sometimes as - 

````
W/System.err: java.lang.ArrayIndexOutOfBoundsException: length=0; index=0
W/System.err:     at org.commcare.utils.EncryptionUtils.decrypt(EncryptionUtils.java:223)
W/System.err:     at org.commcare.utils.EncryptionUtils.decryptFromBase64String(EncryptionUtils.java:268)
W/System.err:     at org.commcare.activities.connect.ConnectIDDatabaseHelper.getConnectDBPassphrase(ConnectIDDatabaseHelper.java:30)
W/System.err:     at org.commcare.activities.connect.ConnectIDDatabaseHelper.init(ConnectIDDatabaseHelper.java:48)
W/System.err:     at org.commcare.activities.connect.ConnectIDManager.init(ConnectIDManager.java:60)
W/System.err:     at org.commcare.activities.LoginActivity.onCreate(LoginActivity.java:125)
````

We probbaly need to either ensure ` r.getEncryptedPassphrase()` can't be null or handle code for it being null

Reply:
I added a null check for the encryptedPassphrase, although I'm confused as to how it ever could have been null since the ConnectKeyRecord would only ever be created with a valid passphrase, i.e. later in the same function.

Reply:
I think `null` check here is more harmful than useful since it will silently eat those errors. I am happy with us removing the `null` check and let the code crash so that we fail fast if we don't see anyway `r.getEncryptedPassphrase()` can be `null`. Also noticed that this whole code block is in a `try catch` with us swallowing the exceptions, wondering if it's better to let app crash in those scenarios as well.

Reply:
The try-catch at this point in the code is to avoid having a bunch of possible exceptions continue bubbling up the call-chain  such that more and more functions have to add a bunch of `throws` to their signatures. I'm now throwing a `RuntimeException` wrapping the original exception in that `catch` block, that way we can still bubble the exception up and let the app crash without having to add a bunch of `throws` elsewhere. Does that work?

Reply:
yes that's perfect.